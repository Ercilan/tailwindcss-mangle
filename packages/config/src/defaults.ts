import process from 'node:process'
import { defaultMangleClassFilter } from '@tailwindcss-mangle/shared'
import { CSS_LANGS_RE } from 'is-css-request'
import type { MangleUserConfig, PatchUserConfig, UserConfig } from './types'

const defaultPipelineInclude = ['**/*.{html,js,ts,jsx,tsx,vue,svelte,astro,elm,php,phtml,mdx,md}', CSS_LANGS_RE]

const defaultPipelineExclude = [/[\\/](node_modules|dist|\.temp|\.cache|\.vscode)[\\/]/]

export function getDefaultPatchConfig(): PatchUserConfig {
  return {
    output: {
      filename: '.tw-patch/tw-class-list.json',
      removeUniversalSelector: true,
      loose: true,
    },
    tailwindcss: {},
  }
}

export function getDefaultMangleUserConfig(): MangleUserConfig {
  return {
    mangleClassFilter: defaultMangleClassFilter,
    include: defaultPipelineInclude,
    exclude: defaultPipelineExclude,
    disabled: process.env.NODE_ENV === 'development',
    classListPath: '.tw-patch/tw-class-list.json',
    classMapOutput: {
      enable: false,
      filename: '.tw-patch/tw-map-list.json',
      loose: true,
    },
    preserveFunction: [],
  }
}

export function getDefaultUserConfig(): UserConfig {
  return {
    patch: getDefaultPatchConfig(),
    mangle: getDefaultMangleUserConfig(),
  }
}
