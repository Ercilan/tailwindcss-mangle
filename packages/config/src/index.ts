export * from './config'
export * from './constants'
export * from './defaults'
export * from './types'
