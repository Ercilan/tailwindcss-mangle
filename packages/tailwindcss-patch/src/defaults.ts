import type { DeepRequired, InternalPatchOptions, PatchOptions } from './types'
import process from 'node:process'
import { defu } from '@tailwindcss-mangle/shared'

export function getDefaultPatchOptions(): DeepRequired<PatchOptions> {
  return {
    applyPatches: {
      exportContext: true,
      extendLengthUnits: false,
    },
    overwrite: true,
  }
}

export function getPatchOptions(options?: PatchOptions) {
  return defu<InternalPatchOptions, Partial<InternalPatchOptions>[]>(
    options,
    {
      basedir: process.cwd(),
    },
    getDefaultPatchOptions(),
  )
}
