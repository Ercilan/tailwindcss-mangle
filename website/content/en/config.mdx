import { Callout } from 'nextra/components'

# 3.Config

The toolchain now exposes a single configuration surface through `@tailwindcss-mangle/config`. Two top-level sections keep responsibilities clear:

- `registry` drives class discovery and Tailwind patching.
- `transformer` controls how the collected classes are rewritten in application code.

Legacy files that still export `patch` / `mangle` continue to load, but new projects should adopt the updated shape for clarity and better tooling.

## Generate a starter file

```sh npm2yarn
npx tw-patch init
```

This command creates a minimal config at the project root:

```ts
import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({})
```

<Callout type="info">
`defineConfig` provides typed completion. Unknown keys are preserved so you can pass additional options directly to `TailwindcssPatcher` or the bundler plugin.
</Callout>

## Example configuration

```ts
import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({
  registry: {
    output: {
      file: '.tw-patch/tw-class-list.json',
      format: 'json',
      pretty: 2,
      stripUniversalSelector: true,
    },
    tailwind: {
      version: 4,
      next: {
        cssEntries: ['dist/tailwind.css'],
        sources: [
          { base: 'src', pattern: '**/*.{astro,tsx,mdx}', negated: false },
        ],
      },
    },
    features: {
      exportContext: true,
      extendLengthUnits: {
        units: ['rpx'],
      },
    },
    cache: {
      enabled: true,
      dir: '.tw-patch/cache',
      strategy: 'merge',
    },
  },
  transformer: {
    sources: {
      include: [/\.(html|vue|tsx|jsx|mdx)$/],
    },
    generator: {
      classPrefix: 'tw-',
      reserveClassName: [/^theme-/],
    },
    registry: {
      file: '.tw-patch/tw-class-list.json',
      mapping: {
        enabled: false,
        file: '.tw-patch/tw-map-list.json',
        loose: true,
      },
    },
    preserve: {
      functions: ['twIgnore'],
    },
  },
})
```

## Registry options

### `output`

- `file` – destination for the extracted inventory (default `.tw-patch/tw-class-list.json`).
- `format` – either `json` or `lines`; the CLI uses JSON by default.
- `pretty` – `false`, a boolean, or a number of spaces when writing JSON.
- `stripUniversalSelector` – skip `*` from the generated list when `true`.

### `tailwind`

- `version` – optional hint (`2`, `3`, or `4`) when multiple versions are installed.
- `legacy` / `classic` – configure Tailwind v2/v3 instances living in workspace packages (`cwd`, `config`).
- `next.cssEntries` – CSS entry files used by the Oxide scanner. Repeat for multi-bundle setups.
- `next.sources` – custom content sources for v4 projects (defaults to `**/*` under `base`).

### `features`

- `exportContext` – expose Tailwind contexts at runtime so the API can read them back later.
- `extendLengthUnits` – customise the built-in length-unit patching; accepts `false`, `true`, or an option object (`units`, `overwrite`, etc.).

### `cache`

- `enabled` – persist collected classes under `node_modules/.cache/tailwindcss-patch` (enabled by default).
- `dir` / `file` – control where the cache set is stored.
- `strategy` – `merge` adds to the existing cache, `overwrite` replaces it on each extraction.

## Transformer options

The transformer block is consumed by both the core runtime and `unplugin-tailwindcss-mangle`.

- `sources.include` / `sources.exclude` – control which files the plugin scans; defaults cover HTML, frameworks, CSS, and MDX.
- `generator` – customise minified names via `classPrefix`, `reserveClassName`, `exclude`, or `customGenerate`.
- `registry.file` – path to the class list produced by the registry.
- `registry.mapping` – `true` reuses defaults, an object enables JSON output (`enabled`, `file`, `loose`), or supply a callback for full control.
- `filter` – override the class-level filter before the generator runs.
- `preserve.functions` – helper names (e.g. `twIgnore`) that mark template literals as safe from mangling.
- `disabled` – when `true`, skips rewriting (defaults to `process.env.NODE_ENV === 'development'`).

<Callout type="info">
Existing configs that expose `patch` / `mangle` still load. The loader converts them to the new shape internally so you can migrate incrementally.
</Callout>

## Using the config

- `tw-patch install` / `tw-patch extract` resolve this file automatically.
- `unplugin-tailwindcss-mangle` mirrors the same configuration, so plugin options and CLI behaviour stay aligned.
- When you need programmatic control, construct a `TailwindcssPatcher` manually and pass the registry portion—the legacy keys are converted for you at runtime.
