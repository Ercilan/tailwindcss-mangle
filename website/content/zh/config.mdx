import { Callout } from 'nextra/components'

# 3.配置

整套工具现在通过 `@tailwindcss-mangle/config` 提供统一的配置入口。顶层包含两个部分：

- `registry` 负责收集 Tailwind 生成的类名并应用补丁。
- `transformer` 则在应用代码中重写这些类名。

旧版仍然接受 `patch` / `mangle` 结构，加载时会被自动转换；新项目建议直接采用更新后的字段，语义更加清晰。

## 快速生成模板

```sh npm2yarn
npx tw-patch init
```

命令会在项目根目录创建一个最小配置：

```ts
import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({})
```

<Callout type="info">
`defineConfig` 提供完善的类型提示。未知字段会被原样保留，方便直接传递给 `TailwindcssPatcher` 或构建插件。
</Callout>

## 配置示例

```ts
import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({
  registry: {
    output: {
      file: '.tw-patch/tw-class-list.json',
      format: 'json',
      pretty: 2,
      stripUniversalSelector: true,
    },
    tailwind: {
      version: 4,
      next: {
        cssEntries: ['dist/tailwind.css'],
        sources: [
          { base: 'src', pattern: '**/*.{astro,tsx,mdx}', negated: false },
        ],
      },
    },
    features: {
      exportContext: true,
      extendLengthUnits: {
        units: ['rpx'],
      },
    },
    cache: {
      enabled: true,
      dir: '.tw-patch/cache',
      strategy: 'merge',
    },
  },
  transformer: {
    sources: {
      include: [/\.(html|vue|tsx|jsx|mdx)$/],
    },
    generator: {
      classPrefix: 'tw-',
      reserveClassName: [/^theme-/],
    },
    registry: {
      file: '.tw-patch/tw-class-list.json',
      mapping: {
        enabled: false,
        file: '.tw-patch/tw-map-list.json',
        loose: true,
      },
    },
    preserve: {
      functions: ['twIgnore'],
    },
  },
})
```

## Registry 选项

### `output`

- `file` – 类名清单的输出位置（默认 `.tw-patch/tw-class-list.json`）。
- `format` – `json` 或 `lines`，CLI 默认写入 JSON。
- `pretty` – 写入 JSON 时的缩进；可以是 `false`、布尔值或数字。
- `stripUniversalSelector` – 为 `true` 时从清单中移除 `*`。

### `tailwind`

- `version` – 可选提示（`2`/`3`/`4`），在多版本共存时用于快速定位。
- `legacy` / `classic` – 配置 v2/v3 所在包的 `cwd` 与 `config`。
- `next.cssEntries` – Tailwind v4 Oxide 扫描使用的 CSS 入口，可重复声明。
- `next.sources` – v4 的内容源，默认为 `base` 下的 `**/*`。

### `features`

- `exportContext` – 将 Tailwind 运行时上下文导出，便于 API 回读。
- `extendLengthUnits` – 自定义长度单位补丁，支持 `false`、`true` 或配置对象（`units`、`overwrite` 等）。

### `cache`

- `enabled` – 是否缓存已收集的类名（默认启用，目录位于 `node_modules/.cache/tailwindcss-patch`）。
- `dir` / `file` – 缓存文件的目录与名称。
- `strategy` – `merge` 追加至现有集合，`overwrite` 每次重新生成。

## Transformer 选项

该模块同时被核心运行时与 `unplugin-tailwindcss-mangle` 使用。

- `sources.include` / `sources.exclude` – 限定需要扫描的文件，默认覆盖 HTML、常见框架、CSS 与 MDX。
- `generator` – 通过 `classPrefix`、`reserveClassName`、`exclude` 或 `customGenerate` 自定义压缩后的类名。
- `registry.file` – 指向 registry 导出的类名清单。
- `registry.mapping` – `true` 复用默认写盘逻辑；对象形式可开启 JSON 输出（`enabled`、`file`、`loose`）；也可以传入回调自行处理。
- `filter` – 在生成前覆写类名过滤函数。
- `preserve.functions` – 指定 helper 名称（如 `twIgnore`），在模板字符串中保留原始内容。
- `disabled` – 为 `true` 时跳过混淆（默认只在 `NODE_ENV=development` 时禁用）。

<Callout type="info">
旧结构（`patch` / `mangle`）依旧兼容，加载时会自动转换成新的字段，迁移可以循序渐进完成。
</Callout>

## 如何使用

- `tw-patch install` / `tw-patch extract` 会自动读取该文件。
- `unplugin-tailwindcss-mangle` 共享同一份配置，CLI 与插件行为保持一致。
- 如果需要编程式调用，直接实例化 `TailwindcssPatcher` 并传入 `registry` 部分即可；旧字段会在运行时完成归一化。
