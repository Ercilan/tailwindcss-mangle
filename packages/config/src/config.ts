import type { TailwindcssMangleConfig } from './types'
import { createDefineConfig, loadConfig } from 'c12'
import fs from 'fs-extra'
import path from 'pathe'
import { CONFIG_NAME } from './constants'
import { getDefaultUserConfig } from './defaults'

export function getConfig(cwd?: string) {
  return loadConfig<TailwindcssMangleConfig>({
    name: CONFIG_NAME,
    defaults: {
      ...getDefaultUserConfig(),
    },
    cwd,
  })
}

export const defineConfig = createDefineConfig<TailwindcssMangleConfig>()

export function initConfig(cwd: string) {
  return fs.outputFile(
    path.resolve(cwd, `${CONFIG_NAME}.config.ts`),
    `import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({
  registry: {
    output: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
  transformer: {
    registry: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
})
`,
    'utf8',
  )
}
