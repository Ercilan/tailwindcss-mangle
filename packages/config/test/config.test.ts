import fs from 'fs-extra'
import os from 'node:os'
import path from 'pathe'
import { defineConfig, getConfig, initConfig } from '@/config'
import { getDefaultUserConfig } from '@/defaults'

describe('config', () => {
  it('init config', async () => {
    const cwd = path.resolve(__dirname, './fixtures/config/initConfig')
    await initConfig(cwd)
    const dest = path.resolve(cwd, 'tailwindcss-mangle.config.ts')
    expect(await fs.readFile(dest, 'utf8')).toMatchSnapshot()
  })

  it('defineConfig helper returns provided config', () => {
    const config = defineConfig({
      patch: {
        output: {
          filename: 'custom.json',
        },
      },
    })

    expect(config).toEqual({
      patch: {
        output: {
          filename: 'custom.json',
        },
      },
    })
  })

  it('getConfig falls back to defaults when config file is absent', async () => {
    const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'config-test-'))
    const { config } = await getConfig(tempDir)
    expect(config).toEqual(getDefaultUserConfig())
    await fs.remove(tempDir)
  })
})
