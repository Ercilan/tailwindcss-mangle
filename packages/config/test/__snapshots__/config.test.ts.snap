// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > init config 1`] = `
"import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({
  registry: {
    output: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
  transformer: {
    registry: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
})
"
`;
